### ISSUE TITLE
CTRL -C event is closing the ConsoleZ window instead of returning to a prompt

### ISSUE THREAD

**Issue by Sheparzo** [cbucher/console#262]
I expect the CTRL -C event to kill the currently running process and return to a bash prompt.  Instead the CTRL -C event is closing the whole window with the latest version of Git for Windows (git version 2.5.0.windows.1).

**Comment by ericblade** [comment]
There's something really weird going on in the new bash and other tools that come with git.  I have that bash as a shell in my ConsoleZ, and the first time I tried to Ctrl-C a command in it, I don't remember what command, it returned me to the bash prompt, then immediately to the cmd.exe prompt.  Which is really weird because there shouldn't -be- a cmd.exe running underneath it, but apparently there is.  And then I pressed enter a couple of times, and discovered that my keyboard input was going to both a bash and cmd.exe that were running in the same window somehow.
Very strange stuff.  I'm not real happy with this git "upgrade".

**Comment by cbucher** [comment]
https://github.com/cbucher/console/wiki/How-to-Report-an-Issue

**Comment by Sheparzo** [comment]
When I try a CTRL C from the hidden console at view console the issue does not occur.  It will just stop the running process and return to a command prompt.
![hidden-console](https://cloud.githubusercontent.com/assets/6540382/9497772/a52a4a9c-4be4-11e5-83d6-f075c1d45de8.png)

When I try CTRL C from git bash directly, the issue does not occur.  It too will just stop the running process and return to the command prompt.
![gitbash](https://cloud.githubusercontent.com/assets/6540382/9497877/0d4a90a0-4be5-11e5-893b-35039c8926d2.png)

The issue only occurs from ConsoleZ and it will close the window (or tab if there are multiple tabs) and it will leave the bash.exe process running.  I sometimes have multiple bash.exe processes left running even after I have hit CTLR C on their ConsoleZ windows and ConsoleZ is no longer running:
![processes](https://cloud.githubusercontent.com/assets/6540382/9498109/22683996-4be6-11e5-9045-0dfde15cebf0.png)

I'm not sure if this a problem with ConsoleZ or with the git bash that gets installed with Git for Windows.  Sorry if I've posted to the wrong place.

**Comment by Sheparzo** [comment]
I'm seeing now that this has already been reported: [issue #230](https://github.com/cbucher/console/issues/230).  There have been several related issues opened in the [git-for-windows](https://github.com/git-for-windows/git) repo.

There are some helpful comments here in [pull request #205](https://github.com/git-for-windows/git/pull/205#issuecomment-133676621) in git-for-windows.

**Comment by mriehema** [comment]
But there is still no solution, right?

**Comment by Sheparzo** [comment]
@mriehema, I am using the following settings and my CTRL-C is working.  I haven't encountered any breaking functionality with this, however @dscho warns that there may be issues with this (https://github.com/git-for-windows/git/pull/205#issuecomment-134962181).

![untitled](https://cloud.githubusercontent.com/assets/6540382/9722593/02c322f8-557e-11e5-9ee0-635703977f8a.png)

![untitled](https://cloud.githubusercontent.com/assets/6540382/9722634/81242958-557e-11e5-9deb-2efefd0c6766.png)

**Comment by mriehema** [comment]
@Sheparzo Thanks! The settings of the Environment was missing in my setup. So CTRL-C is working now, but my aliases in .bashrc, .bash_profile are missing (located in $HOME). :(

**Comment by dscho** [comment]
@mriehema that's because of the issues I commented on in the other thread: hacking the environment variables (i.e. trying to out-guess the compatibility redirector in `/bin/bash.exe`) is never going to work in the long run. You would always have to play catch-up with the settings of that compatibility redirector (such as setting `TERM` and `HOME`, which you forgot here).

As I stated elsewhere: the proper fix will be to find out what *exactly* is going wrong with `/bin/bash.exe` and fixing *that*. (Without breaking everybody else's Git for Windows, of course.)

**Comment by dlong500** [comment]
@cbucher the pull request you made (git-for-windows/git#205) is closed without resolution due to apparent problems.  Is there any more development on this issue that will actually let us use ConsoleZ with the latest Git/bash?

**Comment by dscho** [comment]
> the pull request you made (git-for-windows/git#205) is closed without resolution due to apparent problems.

@dlong500 Yes, it was closed because there was no indication that my concerns would be addressed. In particular, if I had merged that PR, *my* setup would have been broken. If you are interested in this issue, why don't you pick up the ball and open a new Pull Request?

**Comment by dlong500** [comment]
@dscho I certainly am interested in the issue.  However, I don't have the expertise required to fix it myself (nor do I have the time to gain such knowledge).  Wish I did.  That shouldn't mean I can't ask if there is any progress being made considering the issue is a showstopper for anyone wanting to use ConsoleZ with the latest Git/bash versions.

**Comment by KrekkieD** [comment]
Does anyone know the latest version of git-for-windows that does not have this issue?

Also, what is this `TERM=xterm` variable? xterm is not a recognized command when I try to execute that (`$ xterm`), but not sure that's a problem (since `MSYSTEM` is also not executable for me (`$ MINGW64`)).

**Comment by cbucher** [comment]
@KrekkieD There is an easy workaround: configure a **well coded** shell  (which doesn't terminate when receive CTRL-C) in ConsoleZ tab settings.

Try: `cmd.exe /C "git-cmd.exe"`

*git-cmd.exe* will terminate, but ConsoleZ monitors *cmd.exe* and *cmd.exe* will survive.

**Comment by billyc** [comment]
An easy workaround: change your Console shell command to the following:

`C:\Windows\SysWOW64\cmd.exe /c ""C:\Program Files\Git\usr\bin\bash.exe" --login -i"`

**Comment by jeremydyoung** [comment]
I wasn't using ConsoleZ, but Console2 and ran into this same problem. 

The problem was present with:

```
C:\Program Files\Git\bin\sh.exe --login -i
```

But it was NOT present with:

```
C:\Program Files\Git\usr\bin\sh.exe --login -i
```

One of the suggestions above has a screenshot with this difference, but I wanted to provide clarifying feedback that there does appear to be a difference in behavior between bin\sh.exe and usr\bin\sh.exe.

There does not seem to be any difference in behavior between the sh.exe in a folder and the bash.exe in a particular folder.  Namely, bin\bash.exe behaves the same as bin\sh.exe, and usr\bin\bash.exe behaves the same as usr\bin\sh.exe

Hopefully this is useful to someone.

**Comment by billyc** [comment]
@jeremydyoung  i wish there was a +1 button. That's useful.

**Comment by lewins** [comment]
With `C:\Program Files\Git\usr\bin\sh.exe --login -i` my shortcuts didnt worked but it did with `C:\Program Files\Git\usr\bin\bash.exe --login -i`

**Comment by AlexZeitler** [comment]
@jeremydyoung This did the trick! Thanks :+1:

**Comment by mbenson** [comment]
`/usr/bin/sh.exe` avoids the Ctrl-C issue, all right, but seems not to invoke `.bash_profile`. Anybody have a solution to do both?

**Comment by cbucher** [comment]
@mbenson see [here](https://github.com/cbucher/console/issues/262#issuecomment-138506332)

**Comment by yoshiwaan** [comment]
tl;dr for people that are looking to just have a bash environment in Windows using Git for Windows then use the bash.exe version at `C:\Program Files\Git\usr\bin\bash.exe`

I upgraded to the latest git for windows (2.7.1.2-64) and consolez and started to get this problem, and I found out a little more information to add to the comment by @jeremydyoung above

`C:\Program Files\Git\bin\bash.exe` and `C:\Program Files\Git\usr\bin\bash.exe` are flat out not the same thing. The former is 31KB and seems to be basic Mingw64 bash and the latter is 2014KB  and appears to be MSYS2 bash. The Mingw64 version also sets quite a few more environment variables.

C:\Program Files\Git\bin\bash.exe
```
yoshiwaan@local MINGW64 ~
$ uname
MINGW64_NT-6.1
```
C:\Program Files\Git\usr\bin\bash.exe
```
yoshiwaan@local  ~
$ uname
MSYS_NT-6.1
```

Who knows what the actual uses for each are from the Git for Windows perspective, but clearly the two executables are different, have different ctrl+c handling, may have different compilation options and the environment variables also may add different eventualities into the mix, after all this is an environment created for Git for Windows.

For people wanting a more pure bash + git on Windows I would say it's probably worth trying to just install MSYS2 by itself and use pacman (the Arch Linux package manager that MSYS2 uses) to install git. I'm pretty sure that will work but I've not tested. See _general package management_ here: https://sourceforge.net/p/msys2/wiki/MSYS2%20installation/

For those happy to just use the bash in Git for Windows then use the MSYS2 version. If you want to load profiles properly in the same vein as when SSHing onto a remote server then pass the `--login -i` options to bash.exe or sh.exe, which affect the type of shell created. This isn't anything to do with MSYS, Git for Windows or Mingw64, just basics of how the login system of these shells work. See this diagram for more of an idea of how this works : https://shreevatsa.files.wordpress.com/2008/03/bashstartupfiles1.png

**Comment by dscho** [comment]
tl;dr [*You Can't Believe Most Of What You Read On The Internet* (A. Lincoln)](https://en.wikipedia.org/wiki/Abraham_Lincoln)

> for people that are looking to just have a bash environment in Windows using Git for Windows then use the bash.exe version at C:\Program Files\Git\usr\bin\bash.exe

This is flawed advice, as [I pointed out earlier already](https://github.com/cbucher/console/issues/262#issuecomment-138506332).

> The former is 31KB and seems to be basic Mingw64 bash

This is incorrect. `/bin/bash.exe` refers to a redirector that sets up some environment variables and then hands off to `/usr/bin/bash.exe`. This redirector was only invented for backwards compatibility, but apparently is needed for eternity.

To sum it up: `/bin/bash.exe` is *not* the Bash, but relies on `/usr/bin/bash.exe` to do the hard lifting.

**Comment by yoshiwaan** [comment]
@dscho thanks for better explaining about the wrapper. It makes sense.
I don't think my advice is bad as of this time (Git for Windows 2.7.1) because your options _right now_ are:

* Use git-bash.exe and just use the mintty that comes with Git for Windows
* Use `C:\Program Files\Git\bin\bash.exe` and have Ctrl+C kill your window
* Use `C:\Program Files\Git\usr\bin\bash.exe` and have issue due to unset environment variables, which I'm not clearn on what issues will cause other than Vim time being wrong

Of the above if you want to use ConsoleZ then you have to choose the lesser of two evils, of which the latter is better (IMO). Hopefully there's a fix soon for the redirector, more and more these days I wish I knew C. Might be time to bite the bullet.

https://github.com/git-for-windows/git/pull/205 has a lot more info for anyone interested.

**Comment by dscho** [comment]
> your options right now are:
> [...]

You forgot one, rather crucial, option, the one option I would have prefered:

- work with @dscho on a real fix.

All it would need would be a person dedicated to driving the fix. Whatever knowledge is missing, I can provide, but not the time nor the incentive.

> more and more these days I wish I knew C

No one was born with knowledge of C. Chances are that you were exposed to some language similar enough to C, such as Javascript, Ruby, etc. Besides, you only need to imitate the surrounding code. And you are totally allowed to ask questions!

**Comment by BiggerNoise** [comment]
@dscho - I have some (dusty) C knowledge, and can probably take a look at this some time this week.

Can you point me in the direction that you think would be a good place to start? Thanks.

**Comment by dscho** [comment]
@BiggerNoise could you first try with Git for Windows v2.8.1? It might be fixed in the meantime...

**Comment by BiggerNoise** [comment]
Yes, that appears to have done it. My C skills remain safely locked away.

I am starting the shell with "C:\Program Files\Git\bin\bash.exe" and no arguments.
